### PR TITLE
refactor(controller): make read log generate to async and add scanRange impl

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description

1. Make Read log generate process to async.
2. Add scanRange impl for data loader.
3. Fix bug: loss startType in session.
4. Fix bug: loss revison in session.

**TODO**

1. Client consumption use api's isEnclusive to scan records(now client and server use the default value).
2. Use cursor to impl log generation.

## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 1]
```diff
@@ -316,7 +317,170 @@ public RecordList scan(DataStoreScanRequest req) {
...
+                            List<Map<String, Object>> records,
+                            BaseValue lastKey
+                    ) {
+                        var keyColumn = tables.get(0).schema.getKeyColumn();
+                        var ranges = new ArrayList<KeyRangeList.Range>();
+                        var batchSize = req.getRangeInfo().getBatchSize();
+                        var index = 0;
+                        while (index < records.size()) {
+                            if (index + batchSize < records.size()) {
+                                var start = req.isEncodeWithType()
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:378]: It looks like `encodeWithType` can never be false.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 2]
```diff
@@ -482,17 +627,7 @@ record = new HashMap<>();
...
-                table.table.getColumnStatistics(table.columns)
-                        .forEach((k, v) -> columnStatistics.computeIfAbsent(k, x -> new ColumnStatistics()).merge(v));
-            }
-            return new RecordList(columnSchemaMap,
-                    columnStatistics.entrySet().stream()
-                            .collect(Collectors.toMap(Entry::getKey, entry -> entry.getValue().toColumnHintsDesc())),
-                    ret,
-                    (String) BaseValue.encode(lastKey, false, false),
-                    BaseValue.getColumnType(lastKey).name());
+            return resultResolver.apply(tables, columnSchemaMap, ret, lastKey);
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:630]: It does a lot of useless encoding when used by `scanKeyRange`.

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java [hunk 3]
```diff
@@ -106,45 +195,49 @@ public Session generateSession(DataReadRequest request) {
...
             readLog.setConsumerId(consumerId);
             readLog.setAssignedNum(readLog.getAssignedNum() + 1);
             dataReadLogDao.updateToAssigned(readLog);
             log.info("Assignment data id: {} to consumer:{}", readLog.getId(), readLog.getConsumerId());
+        } else {
+            var session = sessionDao.selectOne(sessionId);
+            if (session.getStatus() == Status.SessionStatus.UNFINISHED) {
+                throw new SwRequestFrequentException(
+                        DATASET_LOAD, "data load: index is building, please try again later");
+            }
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java:229]: Is it possible for the following abnormal situation to occur?
1. The async generating was not done (more data should be processed).
2. This function is called and gets nothing.
3. The generating is done and mark the session finished.
4. This function check the session, and see the finished session.

Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java:228]: yes, you are right, already use the origin session object which was passed by outer.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java (hunk 2 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java]: done

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 4]
```diff
@@ -316,7 +318,211 @@ public RecordList scan(DataStoreScanRequest req) {
...
+                                var start = (String) startKey.encode(req.isRawResult(), false);
+                                var startType = startKey.getColumnType().name();
+
+                                var endKey = keys.get(keys.size() - 1);
+                                var end = (String) endKey.encode(req.isRawResult(), false);
+                                var endType = endKey.getColumnType().name();
+                                ranges.add(KeyRangeList.Range.builder()
+                                        .start(start)
+                                        .startType(startType)
+                                        .startInclusive(true)
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:461]: Which part depends on whether the end is specified? 
The resume in `generate`?

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java [hunk 5]
```diff
@@ -68,34 +82,109 @@ public Session getSession(DataReadRequest request) {
...
+        @Override
+        public int compareTo(@NotNull Delayed o) {
+            long diffMillis = getDelay(TimeUnit.MILLISECONDS) - o.getDelay(TimeUnit.MILLISECONDS);
+            diffMillis = Math.min(diffMillis, 1);
+            diffMillis = Math.max(diffMillis, -1);
+            return (int) diffMillis;
+        }
+    }
+
+    @Scheduled(fixedRate = 1000)
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java:141]: fixedDelay ?

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java [hunk 6]
```diff
@@ -68,34 +82,109 @@ public Session getSession(DataReadRequest request) {
...
+    public void dealWithErrorSessions() {
+        while (!failSessionQueue.isEmpty()) {
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java:143]: Dead loop when it is not empty and no element with expired delay?

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java [hunk 7]
```diff
@@ -106,45 +195,48 @@ public Session generateSession(DataReadRequest request) {
                                 .status(Status.DataStatus.UNPROCESSED)
                                 .build())
                         .collect(Collectors.toList()),
-                1000).forEach(dataReadLogDao::batchInsert);
-
-        return session;
-
+                this.insertBatchSize
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java:198]: list may be empty, and the `batchInsert` may fail.

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataStoreIndexProvider.java [hunk 8]
```diff
@@ -44,87 +41,36 @@ public DataStoreIndexProvider(DataStore dataStore) {
...
-                        .end(null)
-                        .endType(keys.get(keys.size() - 1).type)
-                        .size(keys.size() - index)
-                        .build());
-                index = keys.size();
-            }
-        }
-        return indices;
+        var ranges = dataStore.scanKeyRange(DataStoreScanRangeRequest.builder()
+                .start(request.getStart())
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataStoreIndexProvider.java:49]: Miss start type?

Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataStoreIndexProvider.java:49]: done

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java (hunk 7 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java]: done

File: server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java (hunk 6 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java]: fixed

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 9]
```diff
@@ -316,7 +318,211 @@ public RecordList scan(DataStoreScanRequest req) {
...
+                                        .size(batchSize)
+                                        .build());
+                                index += batchSize;
+                            } else {
+                                var startKey = keys.get(index);
+                                var start = (String) startKey.encode(req.isRawResult(), false);
+                                var startType = startKey.getColumnType().name();
+
+                                var endKey = keys.get(keys.size() - 1);
+                                var end = (String) endKey.encode(req.isRawResult(), false);
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java:456]: The encoded key can not use raw result, the `rawResult` is for response record values, not keys.

File: server/controller/src/main/resources/application.yaml [hunk 10]
```diff
@@ -56,6 +56,7 @@ sw:
         log-cache-size: ${DATASET_CONSUMPTION_LOG_CACHE_SIZE:1000}
         log-cache-timeout: ${DATASET_CONSUMPTION_LOG_CACHE_TIMEOUT:24h}
         lock-wait-seconds: ${DATASET_CONSUMPTION_LOCK_WAIT_SECONDS:10}
+        deal-error-session-interval: ${DATASET_CONSUMPTION_DEAL_ERROR_SESSION_INTERNAL:1000}
```
Comment (jialeicui) [inline server/controller/src/main/resources/application.yaml:59]: 1000 to 1s will have better readability? and will give the reader an explicit time format example.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java [hunk 11]
```diff
@@ -316,7 +318,211 @@ public RecordList scan(DataStoreScanRequest req) {
...
+        return scanRecords(req, new ResultResolver<>() {
+                    @Override
+                    public KeyRangeList apply(
+                            List<TableMeta> tables,
+                            Map<String, ColumnSchema> columnSchemaMap,
+                            List<TableRecords> records
+                    ) {
+
+                        BaseValue lastKey;
+                        List<BaseValue> keys = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java]: fixed

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java (hunk 9 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java]: fixed, use false

File: server/controller/src/main/resources/application.yaml (hunk 10 above)
Comment (goldenxinxing) [reply server/controller/src/main/resources/application.yaml]: done

**Review by jialeicui — APPROVED** [review]
LGTM

**Pull request merged**